src/stacktrace.cpp

```diff
@@ -0,0 +1,285 @@
+// stacktrace.h (c) 2008, Timo Bingmann from http://idlebox.net/
+// published under the WTFPL v2.0
+
+// Downloaded from http://panthema.net/2008/0901-stacktrace-demangled/
+// and modified for C++ and FC by Steemit, Inc. and jmjatlanta for the Bitshares Foundation
+
+#include <fc/filesystem.hpp>
+//#include <fc/macros.hpp>
+#include <fc/stacktrace.hpp>
+
+#if defined(__GNUC__) && !defined( __APPLE__ )
+
+#include <cxxabi.h>
+#include <execinfo.h>
+#include <signal.h>
+#include <ucontext.h>
+#include <unistd.h>
+
+#include <cstdlib>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <string>
+
+FILE *popen(const char *command, const char *mode);
+int pclose(FILE *stream);
+
+namespace fc {
+
+//
+// This function is based on http://www.cplusplus.com/forum/beginner/117874/
+// I use something platform-specific and self-contained, rather than using
+// e.g. boost::process, because I want stacktrace to be small and self-contained
+//
+std::string run_command( const std::string& command )
+{
+    FILE* file = popen( command.c_str(), "r" ) ;
+
+    if( file )
+    {
+        std::ostringstream ss;
+
+        constexpr std::size_t MAX_LINE_SZ = 8192;
+        char line[MAX_LINE_SZ];
+
+        while( fgets( line, MAX_LINE_SZ, file ) )
+           ss << line;
+
+        pclose(file);
+        return ss.str();
+    }
+
+    return "";
+}
+
+void print_stacktrace_linenums(std::ostream& out, void** addrlist, int addrlen )
+{
+   if( addrlen == 0 )
+      return;
+
+   char my_path[512];
+   auto result = readlink("/proc/self/exe", my_path, 511);
+   if( result < 0 )
+   {
+      std::cerr << "print_stacktrace_linenums() failed, could not read PID" << std::endl;
+      out << "print_stacktrace_linenums() failed, could not read PID" << std::endl;
+      return;
+   }
+   else
+   {
+      my_path[result] = '\0';
+   }
+
+   std::ostringstream ss_cmd;
+   ss_cmd << "addr2line -p -a -f -C -i -e " << my_path;
```

Comment:
This is a classic shell injection bug. Avoid creating subprocesses using functions that execute a string as a shell command. Use something like execvp instead.